TST, MAINT: torch + GPU failures for test_create_diagonal

On latest `main`, via `SCIPY_DEVICE=cuda python dev.py test -j 32 -b all`, there are 81 new failures similar to the one below the fold with `torch` on device.

<details>

```
__________________________________________________________________________ TestArrayAPI.test_create_diagonal[9-9-torch] ___________________________________________________________________________
[gw0] linux -- Python 3.11.2 /home/treddy/python_venvs/py_311_scipy_dev/bin/python
scipy/_lib/tests/test_array_api.py:201: in test_create_diagonal
    B = xp.asarray(np.diag(x, offset), dtype=one.dtype)
        A          = tensor([[0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
         0.2277, 0.0000, 0.0000, 0.00...000, 0.0000,
         0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000]],
       device='cuda:0')
        n          = 9
        offset     = 9
        one        = tensor(1., device='cuda:0')
        rng        = Generator(PCG64) at 0x7F5616E10900
        self       = <scipy._lib.tests.test_array_api.TestArrayAPI object at 0x7f56b1b7a6d0>
        x          = tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0')
        xp         = <module 'torch' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/torch/__init__.py'>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/numpy/lib/_twodim_base_impl.py:303: in diag
    v = asanyarray(v)
        k          = 9
        v          = tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0')
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/torch/_tensor.py:1081: in __array__
    return handle_torch_function(Tensor.__array__, (self,), self, dtype=dtype)
        dtype      = None
        self       = tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0')
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/torch/overrides.py:1630: in handle_torch_function
    result = mode.__torch_function__(public_api, types, args, kwargs)
        args       = (tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0'),)
        kwargs     = {'dtype': None}
        mode       = <torch.utils._device.DeviceContext object at 0x7f5766ae1210>
        overloaded_args = [tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0')]
        public_api = <function Tensor.__array__ at 0x7f5765d0d8a0>
        relevant_args = (tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0'),)
        types      = (<class 'torch.Tensor'>,)
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/torch/utils/_device.py:79: in __torch_function__
    return func(*args, **kwargs)
        args       = (tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0'),)
        func       = <function Tensor.__array__ at 0x7f5765d0d8a0>
        kwargs     = {'dtype': None}
        self       = <torch.utils._device.DeviceContext object at 0x7f5766ae1210>
        types      = (<class 'torch.Tensor'>,)
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/torch/_tensor.py:1083: in __array__
    return self.numpy()
E   TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
        dtype      = None
        self       = tensor([0.2277, 0.4056, 0.9003, 0.1249, 0.4435, 0.2556, 0.5505, 0.1051, 0.2509],
       device='cuda:0')
```

</details>

The code was recently added in gh-21778.

--- a/scipy/_lib/tests/test_array_api.py
+++ b/scipy/_lib/tests/test_array_api.py
@@ -196,7 +196,7 @@ class TestArrayAPI:
     def test_create_diagonal(self, n, offset, xp):
         rng = np.random.default_rng(2347823)
         one = xp.asarray(1.)
-        x = xp.asarray(rng.random(n), dtype=one.dtype)
-        A = xp_create_diagonal(x, offset=offset, xp=xp)
+        x = rng.random(n)
+        A = xp_create_diagonal(xp.asarray(x, dtype=one.dtype), offset=offset, xp=xp)
         B = xp.asarray(np.diag(x, offset), dtype=one.dtype)
         xp_assert_equal(A, B)

